sessionmanager.md

```diff
@@ -0,0 +1,96 @@
+# `SessionManager` for @azure/service-bus v7
+
+## Introduction
+
+Sessions allow a user to logically subdivide a queue or subscription
+into multiple smaller queues. This is useful for customers that would 
+like to allow for many smaller streams of messages within a single 
+queue or subscription.
+
+## Use case
+
+A simple use case would be having a set of documents. Users can "edit"
+documents, and each edit is stored as a message within Service Bus.
+
+```javascript
+{ 
+  sessionId: "the document name"
+  body: {
+    mode: "edit",
+    content: "content of the edit"
+  }  
+}
+```
+
+Multiple edits or transformations could be sent for a document, at any time.
+
+To handle this the user would need to write code similar to this:
+
+```javascript
+
+  while (true) {    
+    // some sort of check to make sure we're not processing
+    // too many sessions at a time. The odds of having more documents
+    // than available processors is _high_.
+    acquireSemaphore();
+
+    const sessionReceiver: SessionReceiver = getNextUnlockedSessionReceiver();
+    startSessionLockRenewal(sessionReceiver);
+
+    sessionReceiver.subscribe(messageHandlers);
+
+    // ...after some predefined period of time
+    sessionReceiver.close();
+
+    stopSessionLockRenewal(sessionReceiver);
+    releaseSemaphore();
+  }
+```
+
+## Proposal
+
+In the same vein as what we provide in `Receiver.subscribe()` we can provide a 
+`SessionManager` that would support the use case where the user would like 
+to read messages, in parallel, from a number of sessions while controlling
+the amount of parallelism.
+
+This would simplify session usage to the point where it becomes more accessible, 
+allowing more customers to use this powerful feature.
+
+The proposed interface for this change:
+
+```javascript
+class ServiceBusClient {
+  // a session manager is conceptually a receiver, but does not have to 
+  // obey the same interface.
+  createSessionManager("queue", sessionManagerOptions): SessionManager;
+  createSessionManager("topic", "subscription", sessionManagerOptions): SessionManager;
+}
+
+interface SessionManagerOptions {
+  // amount of time we will wait for messages to arrive in a session before
+  // it is `close()`d
+  idleTimeoutMs: number;
+
+  // the maximum number of receivers that can be active at any
+  // time. Allows the customer to avoid getting steamrolled by too many 
+  // sessions.
+  concurrentReceivers: number;
+
+  // TODO: I'd be curious about other potential options here.
+}
+
+interface SessionManager {
+  // the handlers 
+  subscribe(handlers);
```

Comment:
do the handlers need to have a certain signature?

Reply:
They do but it'd be the same as the one used for a general `Receiver`, no difference.

Reply:
Any chance you can link that here?

Reply:
https://github.com/Azure/azure-sdk-for-js/blob/47cfb37503443bfcdd16c00665214f48e0d9dd92/sdk/servicebus/service-bus/src/models.ts#L11

Reply:
So I believe the settlement operations are on the message. But what about session level operations? Is there a way to do those from the handler without adding the Receiver as a parameter?

Reply:
Chatting with @KieranBrantnerMagee in other threads - I'm starting to think it might be better to consider this a brand new interface. There are some cases I'm thinking that would be nice to avoid:

- If the user wants to set state on a session how do we support it? We could pass them a session context as part of their subscribe() handler. However, that's not _great_ when we're dealing with non-subscribe() methods. So perhaps we should just drop those.
- Some options don't really make sense anymore - like setting auto-lock-renewal for the messages.

Reply:
Yeah for session state, I believe you would need to expose this method.
For max-auto-lock-renewal, I think this would still apply potentially, but it would be for the session. This is what we do in .NET - when using the session processor, max auto lock renewal applies to the session.

Reply:
I've added a section and revised this one. I think you made enough good points that I decided to just remove the compatibility requirement.